File: src/core/analyzer.ts
import type { PromptSegment, PromptAnalysis, Strategy } from '../types.js'
import { estimateTokens, estimateTokensForObject } from './token-estimator.js'
import { ContentDiffer } from './differ.js'

/**
 * The heart of promptcache. Analyzes a prompt to identify which parts
 * are worth caching and where to place breakpoints.
 *
 * Scoring is based on:
 * - Position: system prompts and tools are almost always static
 * - Size: bigger blocks save more money when cached
 * - Repetition: if we've seen it before, it's definitely stable
 * - Content patterns: XML tags, JSON schemas, large text = likely static
 */
export class PromptAnalyzer {
  private differ: ContentDiffer
  private tokenCounter: (text: string) => number

  constructor(
    differ?: ContentDiffer,
    tokenCounter?: (text: string) => number,
  ) {
    this.differ = differ || new ContentDiffer()
    this.tokenCounter = tokenCounter || estimateTokens
  }

  analyzeAnthropicParams(params: {
    system?: string | Array<{ type: string; text?: string; [key: string]: unknown }>
    messages?: Array<{ role: string; content: string | Array<{ type: string; text?: string; [key: string]: unknown }> }>
    tools?: Array<Record<string, unknown>>
  }): PromptAnalysis {
    const segments: PromptSegment[] = []

    // analyze system prompt
    if (params.system) {
      if (typeof params.system === 'string') {
        const tokens = this.tokenCounter(params.system)
        const repetitions = this.differ.recordAndCount(params.system)
        segments.push({
          path: 'system',
          content: params.system.slice(0, 200),
          tokenEstimate: tokens,
          stabilityScore: this.scoreStability('system', repetitions, params.system),
          type: 'system',
        })
      } else if (Array.isArray(params.system)) {
        params.system.forEach((block, i) => {
          if (block.type === 'text' && block.text) {
            const tokens = this.tokenCounter(block.text)
            const repetitions = this.differ.recordAndCount(block.text)
            segments.push({
              path: `system[${i}]`,
              content: block.text.slice(0, 200),
              tokenEstimate: tokens,
              stabilityScore: this.scoreStability('system', repetitions, block.text),
              type: 'system',
            })
          }
        })
      }
    }

    // analyze tools
    if (params.tools && params.tools.length > 0) {
      const toolsJson = JSON.stringify(params.tools)
      const tokens = estimateTokensForObject(params.tools)
      const repetitions = this.differ.recordObjectAndCount(params.tools)
      segments.push({
        path: 'tools',
        content: `${params.tools.length} tool definitions`,
        tokenEstimate: tokens,
        stabilityScore: this.scoreStability('tool', repetitions),
        type: 'tool',
      })
    }

    // analyze messages
    if (params.messages) {
      const totalMessages = params.messages.length
      params.messages.forEach((msg, i) => {
        const text = typeof msg.content === 'string'
          ? msg.content
          : JSON.stringify(msg.content)
        const tokens = this.tokenCounter(text)

        // messages near the end of the conversation are likely the variable part
        const positionFromEnd = totalMessages - i
        const repetitions = this.differ.recordAndCount(text)

        let baseScore: number
        if (positionFromEnd <= 1) {
          baseScore = 0.1 // current turn — always changing
        } else if (positionFromEnd <= 3) {
          baseScore = 0.3 // recent turns — probably changing
        } else {
          baseScore = 0.7 // older turns — likely stable in a session
        }

        segments.push({
          path: `messages[${i}]`,
          content: text.slice(0, 200),
          tokenEstimate: tokens,
          stabilityScore: this.scoreStability('message', repetitions, text, baseScore),
          type: 'message',
        })
      })
    }

    // build the analysis
    const stableSegments = segments.filter(s => s.stabilityScore >= 0.6)
    const variableSegments = segments.filter(s => s.stabilityScore < 0.6)
    const totalTokens = segments.reduce((sum, s) => sum + s.tokenEstimate, 0)
    const cacheableTokens = stableSegments.reduce((sum, s) => sum + s.tokenEstimate, 0)

    return {
      segments,
      stableSegments,
      variableSegments,
      totalTokens,
      cacheableTokens,
      estimatedSavingsPercent: totalTokens > 0
        ? Math.round((cacheableTokens / totalTokens) * 90) // 90% is the cache read discount
        : 0,
    }
  }

  private scoreStability(
    type: string,
    repetitions: number,
    content?: string,
    baseScore?: number,
  ): number {
    let score = baseScore ?? 0.5

    // structural position matters a lot
    if (type === 'system') score = Math.max(score, 0.9)
    if (type === 'tool') score = Math.max(score, 0.9)

    // seen before? that's a strong signal
    if (repetitions >= 3) score = Math.max(score, 0.95)
    else if (repetitions >= 2) score = Math.max(score, 0.85)

    // content heuristics
    if (content) {
      // XML-style tags usually indicate structured static instructions
      if (content.includes('<instructions>') || content.includes('<context>') || content.includes('<rules>')) {
        score = Math.max(score, 0.85)
      }
      // large text blocks are usually documents/knowledge bases
      if (content.length > 5000) {
        score = Math.max(score, 0.8)
      }
      // JSON schemas
      if (content.includes('"type"') && content.includes('"properties"')) {
        score = Math.max(score, 0.85)
      }
    }

    return Math.min(score, 1)
  }
}

File: python/cachellm/core/analyzer.py
from __future__ import annotations

import json
from typing import Any, Callable, Dict, List, Optional

from cachellm.types import PromptAnalysis, PromptSegment
from cachellm.core.token_estimator import estimate_tokens, estimate_tokens_for_object
from cachellm.core.differ import ContentDiffer


class PromptAnalyzer:
    """The heart of cachellm. Analyzes a prompt to identify which parts
    are worth caching and where to place breakpoints.

    Scoring is based on:
    - Position: system prompts and tools are almost always static
    - Size: bigger blocks save more money when cached
    - Repetition: if we've seen it before, it's definitely stable
    - Content patterns: XML tags, JSON schemas, large text = likely static
    """

    def __init__(
        self,
        differ: Optional[ContentDiffer] = None,
        token_counter: Optional[Callable[[str], int]] = None,
    ):
        self._differ = differ or ContentDiffer()
        self._token_counter = token_counter or estimate_tokens

    def analyze_anthropic_params(self, params: Dict[str, Any]) -> PromptAnalysis:
        segments: List[PromptSegment] = []

        # analyze system prompt
        system = params.get("system")
        if system:
            if isinstance(system, str):
                tokens = self._token_counter(system)
                repetitions = self._differ.record_and_count(system)
                segments.append(PromptSegment(
                    path="system",
                    content=system[:200],
                    token_estimate=tokens,
                    stability_score=self._score_stability("system", repetitions, system),
                    type="system",
                ))
            elif isinstance(system, list):
                for i, block in enumerate(system):
                    if block.get("type") == "text" and block.get("text"):
                        text = block["text"]
                        tokens = self._token_counter(text)
                        repetitions = self._differ.record_and_count(text)
                        segments.append(PromptSegment(
                            path=f"system[{i}]",
                            content=text[:200],
                            token_estimate=tokens,
                            stability_score=self._score_stability("system", repetitions, text),
                            type="system",
                        ))

        # analyze tools
        tools = params.get("tools")
        if tools and len(tools) > 0:
            tokens = estimate_tokens_for_object(tools)
            repetitions = self._differ.record_object_and_count(tools)
            segments.append(PromptSegment(
                path="tools",
                content=f"{len(tools)} tool definitions",
                token_estimate=tokens,
                stability_score=self._score_stability("tool", repetitions),
                type="tool",
            ))

        # analyze messages
        messages = params.get("messages")
        if messages:
            total_messages = len(messages)
            for i, msg in enumerate(messages):
                content = msg.get("content", "")
                text = content if isinstance(content, str) else json.dumps(content)
                tokens = self._token_counter(text)

                position_from_end = total_messages - i
                repetitions = self._differ.record_and_count(text)

                if position_from_end <= 1:
                    base_score = 0.1
                elif position_from_end <= 3:
                    base_score = 0.3
                else:
                    base_score = 0.7

                segments.append(PromptSegment(
                    path=f"messages[{i}]",
                    content=text[:200],
                    token_estimate=tokens,
                    stability_score=self._score_stability("message", repetitions, text, base_score),
                    type="message",
                ))

        stable_segments = [s for s in segments if s.stability_score >= 0.6]
        variable_segments = [s for s in segments if s.stability_score < 0.6]
        total_tokens = sum(s.token_estimate for s in segments)
        cacheable_tokens = sum(s.token_estimate for s in stable_segments)

        return PromptAnalysis(
            segments=segments,
            stable_segments=stable_segments,
            variable_segments=variable_segments,
            total_tokens=total_tokens,
            cacheable_tokens=cacheable_tokens,
            estimated_savings_percent=(
                round((cacheable_tokens / total_tokens) * 90)
                if total_tokens > 0
                else 0
            ),
        )

    def _score_stability(
        self,
        type: str,
        repetitions: int,
        content: Optional[str] = None,
        base_score: Optional[float] = None,
    ) -> float:
        score = base_score if base_score is not None else 0.5

        # structural position matters a lot
        if type == "system":
            score = max(score, 0.9)
        if type == "tool":
            score = max(score, 0.9)

        # seen before? that's a strong signal
        if repetitions >= 3:
            score = max(score, 0.95)
        elif repetitions >= 2:
            score = max(score, 0.85)

        # content heuristics
        if content:
            if "<instructions>" in content or "<context>" in content or "<rules>" in content:
                score = max(score, 0.85)
            if len(content) > 5000:
                score = max(score, 0.8)
            if '"type"' in content and '"properties"' in content:
                score = max(score, 0.85)

        return min(score, 1.0)
